src/wast-parser.cc

```diff
@@ -1222,9 +1224,68 @@ Result WastParser::ParseScript(std::unique_ptr<Script>* out_script) {
   }
 }
 
+Result WastParser::ParseCustomSectionAnnotation(Module* module) {
+  WABT_TRACE(ParseCustomSectionAnnotation);
+  Location loc = GetLocation();
+  Token token = Consume();
+  if (token.text() != "custom") {
+    assert(
+        !"ParseCustomSectionAnnotation should only be called if PeekIsCustom() is true");
+    return Result::Error;
+  }
+  std::string section_name;
+  CHECK_RESULT(ParseQuotedText(&section_name));
+  if (Match(TokenType::Lpar)) {
+    if (!PeekMatch(TokenType::After) && !PeekMatch(TokenType::Before)) {
+      return ErrorExpected({"before", "after"});
+    }
+    Consume();
+    switch (Peek()) {
+      case TokenType::Function:
+      case TokenType::Type:
+      case TokenType::Import:
+      case TokenType::Export:
+      case TokenType::Table:
+      case TokenType::Global:
+      case TokenType::Elem:
+      case TokenType::Data:
+      case TokenType::Memory:
+      case TokenType::Code:
+      case TokenType::Start: {
+        Consume();
+        break;
+      }
+      default: {
+        return ErrorExpected({"type", "import", "function", "table", "memory",
+                              "global", "export", "start", "elem", "code",
+                              "data"});
+      }
+    }
+    EXPECT(Rpar);
+  }
+  std::vector<uint8_t> data;
+  CHECK_RESULT(ParseTextList(&data));
+  EXPECT(Rpar);
+
+  Custom custom = Custom(loc, section_name, data);
+  module->customs.push_back(custom);
+
+  return Result::Ok;
+}
+
+bool WastParser::PeekIsCustom() {
+  // If IsLparAnn succeeds, tokens_.front() must have text, as it is an LparAnn
+  // token.
+  return IsLparAnn(PeekPair()) && tokens_.front().text() == "custom";
```

Comment:
Can we throw a `options_->features.annotations_enabled() &&` into this test for safety (and consistency with `PeekMatchRefType`)?